Overhaul decoration traits and implement isometric selection boxes

pchote opened the pull request:
<img width="620" alt="Screenshot 2020-03-17 at 16 01 31" src="https://user-images.githubusercontent.com/167819/76875568-a0177200-6868-11ea-8d75-c03764fee328.png">

This PR has two major parts:
* Giving the decoration traits their long-delayed overhaul, removing accrued technical debt and reducing a lot of unnecessary duplication (first 6 commits). This also allows mods to define their own selection boxes, and removes the assumptions about selection boxes being rectangular.
* Introducing hexagonal selection boxes for TS structures, both for rendering and for mouse interaction. The selection box heights are automatically ported from the original art.ini (see the `CopyIsometricSelectableHeight` update rule) - some of the original sizes are weird, but i'd prefer we deferred polishing these to a followup PR.

Fixes #4783.
Closes #17346.

I strongly recommend reviewing by commit to keep this manageable.

pchote committed:
Remove IEquatable from ActorBoundsPair.

pchote committed:
Add support for polygon selection shapes.

pchote committed:
Move Interactable and Selectable to Mods.Common.

pchote committed:
Rework decoration renderable traits:

- Removed implicit pip definitions and IPips interface.
  New decoration traits have been added to render them.
  Pip types are no longer hardcoded in OpenRA.Game.

- Decoration rendering is now managed by SelectionDecorations(Base),
  which allows us to remove assumptions about the selection box
  geometry from the decoration traits.

- RenderNameTag has been replaced by WithNameTagDecoration, which is
  an otherwise normal decoration trait.

- Unify the configuration and reduce duplication between traits.

- Removed hardcoded references to specific selection box renderables.

- Remove legacy cruft.

pchote committed:
Update TD decorations.

pchote committed:
Update TS decorations.

pchote committed:
Update D2k decorations.

pchote committed:
Update RA decorations.

abcdefg30 commented on OpenRA.Game/Primitives/Polygon.cs:
```diff
@@ -0,0 +1,111 @@
+
+		public override int GetHashCode()
+		{
+			// TODO: Account for polygon
```
Don't we want to resolve this?

abcdefg30 commented on OpenRA.Game/Exts.cs:
```diff
@@ -101,6 +101,16 @@ public static bool PolygonContains(this int2[] polygon, int2 p)
 
+		public static bool LinesIntersect(int2 a, int2 b, int2 c, int2 d)
+		{
+			// If line segments AB and BC intersect:
```
This comment doesn't make much sense to me. `AB` and `BC` should intersect by definition because they share point `B`?

abcdefg30 commented on OpenRA.Mods.Common/Traits/IsometricSelectable.cs:
```diff
@@ -0,0 +1,114 @@
+		public readonly SelectionPriorityModifiers PriorityModifiers = SelectionPriorityModifiers.None;
+
+		[Desc("All units having the same selection class specified will be selected with select-by-type commands (e.g. double-click). "
+		+ "Defaults to the actor name when not defined or inherited.")]
```
This should probably use `,` instead of `+` to be in line with the rest above.

Mailaender commented on OpenRA.Mods.Common/Traits/Render/WithAmmoPipsDecoration.cs:
```diff
@@ -0,0 +1,84 @@
+		public readonly string EmptySequence = "pip-empty";
+
+		[SequenceReference("Image")]
+		[Desc("Sequence used for full pips that aren't defined in ResourceSequences.")]
```
Copy pasta from elsewhere.

Mailaender commented on OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs:
```diff
@@ -0,0 +1,103 @@
+		public readonly string EmptySequence = "pip-empty";
+
+		[SequenceReference("Image")]
+		[Desc("Sequence used for full pips that aren't defined in ResourceSequences.")]
```
Same here.

abcdefg30 commented on OpenRA.Mods.Common/UpdateRules/Rules/AddPipDecorationTraits.cs:
```diff
@@ -0,0 +1,273 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
2020.

abcdefg30 commented on OpenRA.Mods.Common/UpdateRules/Rules/CopyIsometricSelectableHeight.cs:
```diff
@@ -0,0 +1,86 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
2020.

abcdefg30 commented on OpenRA.Mods.Common/UpdateRules/Rules/ModernizeDecorationTraits.cs:
```diff
@@ -0,0 +1,123 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
2020.

Mailaender commented on OpenRA.Mods.Common/UpdateRules/Rules/CopyIsometricSelectableHeight.cs:
```diff
@@ -0,0 +1,86 @@
+					if (!section.Contains("Height"))
+						continue;
+
+					selectionHeight[section.Name] = (int)(float.Parse(section.GetValue("Height", "1")) * 24);
```
RA2 will need 30 so maybe not hard-code it, but parse from `modData`?

pchote commented:
Fixed.

Mailaender commented on OpenRA.Mods.Common/UpdateRules/Rules/CopyIsometricSelectableHeight.cs:
```diff
@@ -0,0 +1,87 @@
+					if (!section.Contains("Height"))
+						continue;
+
+					selectionHeight[section.Name] = (int)(float.Parse(section.GetValue("Height", "1")) * grid.TileSize.Height);
```
```suggestion
					selectionHeight[section.Name] = (int)(float.Parse(section.GetValue("Height", "1")) * grid.TileSize.Height / 2);
```

Mailaender replied:
Actually, please ignore this.

Mailaender reviewed (dismissed):
Looks great and I also can't spot regressions in the legacy mods.

teinarss commented:
Found this when testing, while a building is repairing the bar moves back and forth. I was at zoomed in at maximum.
![isobars](https://user-images.githubusercontent.com/19173272/77117161-c0465d00-6a31-11ea-83ac-fac7a7ac2bf1.gif)

pchote commented:
I can't reproduce the issue. Can you provide a more specific repro case (including system display dpi, ui scale, and battlefield viewport setting)?

teinarss commented:
![image](https://user-images.githubusercontent.com/19173272/77146658-da148e00-6a8b-11ea-9d7d-1c0261ba1a77.png)

dragunoff commented:
I also got a repro of the moving health bar. Actually happens with any bar that's animating (production bars for example). Happens only at 100% UI scale and at any zoom level. Tested on Ubuntu 19.10.

pchote commented:
I still can't reproduce this, even at 100% UI scale on my 1440p non-retina display (which should be matching your setup above?).

I've changed the calculation for the cut point between full and empty bars to round to the nearest pixel, which was probably the cause of the issue. I also fixed the light/dark colours being inverted on the applied damage bit of the bar.

Can you please test again?

teinarss commented:
> 
> 
> I still can't reproduce this, even at 100% UI scale on my 1440p non-retina display (which should be matching your setup above?).
> 
> I've changed the calculation for the cut point between full and empty bars to round to the nearest pixel, which was probably the cause of the issue. I also fixed the light/dark colours being inverted on the applied damage bit of the bar.
> 
> Can you please test again?

Works!

TheChosenEvilOne commented:
Feels fine to me and I don't see any bars move around.

But in the original game the building selection was done by clicking the tiles and not the building itself, is it possible to reproduce this behaviour?

And as a bonus:
Original TS
![image](https://user-images.githubusercontent.com/34602646/77226744-56d66380-6b83-11ea-9ae6-ab0517dee3fb.png)
OpenRA TS
![image](https://user-images.githubusercontent.com/34602646/77226900-7457fd00-6b84-11ea-8f63-0aa2e6e4f8d7.png)

pchote commented:
> But in the original game the building selection was done by clicking the tiles and not the building itself, is it possible to reproduce this behaviour?

Updated to add `DecorationBounds` and `DecorationHeight` fields. A "TS Classic" style mod can now define `Height: 0` and e.g. `DecorationHeight: 48` to reproduce the legacy behaviour. This feels like a step back in usability IMO, so I don't want to enable this in the default mod unless we have a very good reason to need it.

Also fixed the voice support for isometric selectables (will eventually be needed by the mobile war factories) by adding `Voice` to the new `ISelectableInfo` interface.

TheChosenEvilOne commented:
> > But in the original game the building selection was done by clicking the tiles and not the building itself, is it possible to reproduce this behaviour?
> 
> Updated to add `DecorationBounds` and `DecorationHeight` fields. A "TS Classic" style mod can now define `Height: 0` and e.g. `DecorationHeight: 48` to reproduce the legacy behaviour. This feels like a step back in usability IMO so I don't want to enable this in the default mod unless we have a very good reason to.
> 
> Also fixed the voice support for isometric selectables (will eventually be needed by the mobile war factories) by adding `Voice` to the new `ISelectableInfo` interface.

There are uses for this such as when you want to move units behind buildings but end up either selecting the building or attacking it. I know pressing alt allows you to move behind buildings but it is something you might not remember to press when playing.

pchote commented:
I think it remains to be seen whether the cells-behind-building issue is a problem in practice: I haven't seen any complaints about the selection overlap being a game breaking issue in SP or RV, and this PR improves the situation massively by cutting the corners off the previously square selection boxes. Splitting the selection and decoration bounds breaks the behaviour of the rollover health bar.

TheChosenEvilOne commented:
Show on damage for status bars isn't working correctly, in TS all units have status bars always visible and in RA buildings too have them visible always.
![image](https://user-images.githubusercontent.com/34602646/77229934-62cd2000-6b99-11ea-9df6-aa5475cd1996.png)
![image](https://user-images.githubusercontent.com/34602646/77229938-65c81080-6b99-11ea-8585-4da074799020.png)
![image](https://user-images.githubusercontent.com/34602646/77229969-97d97280-6b99-11ea-83c0-967adbc8cfa0.png)

pchote committed:
Implement isometric selection boxes for TS structures.

pchote commented:
Fixed.

dragunoff commented:
No more stuttering bars - tested on Linux and Win7. However I noticed something else (which is also present on release): there seems to be a duplicate decoration drawn on rollover for unselected buildings (I'm judging by the darkening of the semi-transparent part of the bar). See the screen capture for illustration:
![uta2jA7MHg](https://user-images.githubusercontent.com/1355810/77237136-3384d600-6bce-11ea-89e5-7c24f44552a6.gif)

pchote committed:
Remove special-case rollover rendering.

pchote commented:
Added a new commit to fix that.

atlimit8 commented on OpenRA.Game/Primitives/Polygon.cs:
```diff
@@ -0,0 +1,114 @@
+		public static readonly Polygon Empty = new Polygon(Rectangle.Empty);
+
+		public readonly Rectangle BoundingRect;
+		public readonly int2[] Vertices;
```
I believe it is best to leave this as is for performance. I just want to note that code will have to be audited for changes to vertices after construction, since the bounding rectangle and rectangle and rectangle flag would fall out of sync.

pchote replied:
Maybe i'm missing something, but both `Rectangle` and `int2` are immutable so this shouldn't be possible?

Edit: ah I see, by replacing the array element. If someone intentionally does this then they deserve to have things break :D

atlimit8 commented on OpenRA.Game/SelectableExts.cs:
```diff
@@ -73,14 +73,15 @@ public static FrozenActor WithHighestSelectionPriority(this IEnumerable<FrozenAc
-				bounds.Left + bounds.Size.Width / 2,
-				bounds.Top + bounds.Size.Height / 2);
+				br.Left + br.Size.Width / 2,
+				br.Top + br.Size.Height / 2);
```
Just noting that the centroid need not be the same as the bounding box center. An example is a regular pentagon, which is also an example of radial symmetry since all regular polygons have rotational symmetry. Polygons with rotational symmetry of a degree divisible by four and polygons with both horizontal and vertical symmetry such as the isometric and dimetric projections of a cube do have the center of their bounding boxes as their centroids. The TS mod w is an example. Modders should avoid shapes that deviate too much from those two sets since that would cause irregularities in selection edge cases.

atlimit8 commented on OpenRA.Mods.Common/Traits/Render/SelectionDecorationsBase.cs:
```diff
@@ -0,0 +1,146 @@
+			// Extra bars are shown when:
+			//  * actor is selected
+			//  * status bar preference is set to "always show"
+			//  * status bar preference is set to "when damaged"
```
As discussed in the IRC channel, `//  * status bar preference is set to "when damaged"` is copy pasta. Either the two lines above should be replaced with something like `//  * status bar preference is set to "always show" or "when damaged"`, or the implementation should be changed. I have no preference here, but changing the comment would be the easiest.

atlimit8 commented on OpenRA.Mods.Common/Traits/Render/WithDecorationBase.cs:
```diff
@@ -0,0 +1,147 @@
+		}
+	}
+
+	public abstract class WithDecorationBase : ConditionalTrait<WithDecorationBaseInfo>, IDecoration
```
Duplicate references to trait info in sub-classes can be removed by continuing the curiously recurring template pattern I gave `ConditionalTrait<Info>`
```c#
public abstract class WithDecorationBase<InfoType> : ConditionalTrait<Info>, IDecoration where InfoType : WithDecorationBaseInfo
```

pchote commented:
Rebased (first push) and fixed (second push).

atlimit8 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/660cded43de8245cca1b788ab62f0b66aa242ec5)

sorcerer86pt commented:
![Base Profile Screenshot 2020 03 28 - 19 23 29 03](https://user-images.githubusercontent.com/11853415/77831871-a68bd080-7129-11ea-9465-b2c53796f99f.png)
![Base Profile Screenshot 2020 03 28 - 19 23 40 34](https://user-images.githubusercontent.com/11853415/77831873-a7bcfd80-7129-11ea-839a-0cda87a3893e.png)
